File: sdks/python/tests/integration/helpers/mock_server.py
"""Mock gRPC server for integration testing."""

import asyncio
from typing import Dict, Optional
from dataclasses import dataclass, field
from datetime import datetime

import grpc
from grpc import aio

from norikv.proto import (
    KvServicer,
    MetaServicer,
    PutRequest,
    PutResponse,
    GetRequest,
    GetResponse,
    DeleteRequest,
    DeleteResponse,
    Version as ProtoVersion,
)


@dataclass
class StoredValue:
    """Value stored in mock server."""

    value: bytes
    version: ProtoVersion
    ttl_ms: int = 0
    created_at: datetime = field(default_factory=datetime.now)


class MockKvServicer(KvServicer):
    """Mock implementation of KV service."""

    def __init__(self):
        """Initialize the mock KV servicer."""
        self._store: Dict[bytes, StoredValue] = {}
        self._version_counter = 0
        self._is_leader = True
        self._leader_hint: Optional[str] = None

    def set_leader_status(self, is_leader: bool, leader_hint: Optional[str] = None):
        """Set whether this node is the leader."""
        self._is_leader = is_leader
        self._leader_hint = leader_hint

    def clear(self):
        """Clear all stored data."""
        self._store.clear()
        self._version_counter = 0

    def _next_version(self) -> ProtoVersion:
        """Generate next version."""
        self._version_counter += 1
        return ProtoVersion(term=1, index=self._version_counter)

    async def Put(self, request: PutRequest, context: grpc.aio.ServicerContext) -> PutResponse:
        """Handle Put request."""
        # Check if we're the leader
        if not self._is_leader:
            await context.abort(
                grpc.StatusCode.FAILED_PRECONDITION,
                f"NOT_LEADER: {self._leader_hint or 'unknown'}",
            )

        # Check if_not_exists
        if request.HasField("if_not_exists") and request.key in self._store:
            await context.abort(
                grpc.StatusCode.ALREADY_EXISTS,
                "Key already exists",
            )

        # Check if_match version
        if request.HasField("if_match"):
            stored = self._store.get(request.key)
            if not stored:
                await context.abort(
                    grpc.StatusCode.NOT_FOUND,
                    "Key not found for conditional update",
                )
            if (
                stored.version.term != request.if_match.term
                or stored.version.index != request.if_match.index
            ):
                await context.abort(
                    grpc.StatusCode.FAILED_PRECONDITION,
                    "Version mismatch",
                )

        # Store the value
        version = self._next_version()
        self._store[request.key] = StoredValue(
            value=request.value,
            version=version,
            ttl_ms=request.ttl_ms,
        )

        return PutResponse(version=version)

    async def Get(self, request: GetRequest, context: grpc.aio.ServicerContext) -> GetResponse:
        """Handle Get request."""
        stored = self._store.get(request.key)
        if not stored:
            return GetResponse()

        # Check TTL expiration
        if stored.ttl_ms > 0:
            elapsed_ms = (datetime.now() - stored.created_at).total_seconds() * 1000
            if elapsed_ms > stored.ttl_ms:
                # Expired - remove and return not found
                del self._store[request.key]
                return GetResponse()

        return GetResponse(
            value=stored.value,
            version=stored.version,
        )

    async def Delete(
        self, request: DeleteRequest, context: grpc.aio.ServicerContext
    ) -> DeleteResponse:
        """Handle Delete request."""
        # Check if we're the leader
        if not self._is_leader:
            await context.abort(
                grpc.StatusCode.FAILED_PRECONDITION,
                f"NOT_LEADER: {self._leader_hint or 'unknown'}",
            )

        # Check if_match version
        if request.HasField("if_match"):
            stored = self._store.get(request.key)
            if not stored:
                return DeleteResponse(tombstoned=False)
            if (
                stored.version.term != request.if_match.term
                or stored.version.index != request.if_match.index
            ):
                await context.abort(
                    grpc.StatusCode.FAILED_PRECONDITION,
                    "Version mismatch",
                )

        # Delete the key
        if request.key in self._store:
            del self._store[request.key]
            return DeleteResponse(tombstoned=True)

        return DeleteResponse(tombstoned=False)


class MockMetaServicer(MetaServicer):
    """Mock implementation of Meta service."""

    def __init__(self):
        """Initialize the mock Meta servicer."""
        self._is_healthy = True

    def set_health(self, is_healthy: bool):
        """Set the health status."""
        self._is_healthy = is_healthy


class MockServer:
    """Mock gRPC server for testing."""

    def __init__(self, port: int = 0):
        """Initialize the mock server.

        Args:
            port: Port to listen on (0 for random available port)
        """
        self._port = port
        self._server: Optional[aio.Server] = None
        self._kv_servicer = MockKvServicer()
        self._meta_servicer = MockMetaServicer()

    async def start(self) -> int:
        """Start the mock server.

        Returns:
            The port the server is listening on
        """
        from norikv.proto import add_KvServicer_to_server, add_MetaServicer_to_server

        self._server = aio.server()
        add_KvServicer_to_server(self._kv_servicer, self._server)
        add_MetaServicer_to_server(self._meta_servicer, self._server)

        port = self._server.add_insecure_port(f"[::]:{self._port}")
        await self._server.start()

        return port

    async def stop(self):
        """Stop the mock server."""
        if self._server:
            await self._server.stop(grace=1.0)
            self._server = None

    def get_kv_servicer(self) -> MockKvServicer:
        """Get the KV servicer for test control."""
        return self._kv_servicer

    def get_meta_servicer(self) -> MockMetaServicer:
        """Get the Meta servicer for test control."""
        return self._meta_servicer

    async def __aenter__(self) -> "MockServer":
        """Async context manager entry."""
        await self.start()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit."""
        await self.stop()

File: sdks/typescript/tests/integration/helpers/mock-server.ts
/**
 * Mock gRPC server for integration testing.
 *
 * Provides an in-memory implementation of NoriKV gRPC services for testing
 * SDK behavior without requiring a real server.
 */

import * as grpc from '@grpc/grpc-js';
import { KvService, MetaService, type PutRequest, type PutResponse, type GetRequest, type GetResponse, type DeleteRequest, type DeleteResponse, type ClusterView } from '@norikv/client/proto/norikv';

interface MockKVStore {
  [key: string]: {
    value: Buffer;
    version: { term: number; index: number };
    ttl?: number;
    createdAt: number;
  };
}

interface CallRecord {
  method: string;
  request: any;
  timestamp: number;
}

export class MockNoriKVServer {
  private server: grpc.Server;
  private store: MockKVStore = {};
  private callHistory: CallRecord[] = [];
  private port: number = 0;
  private currentTerm: number = 1;
  private currentIndex: number = 0;
  private leaderId: string = 'mock-leader';
  private notLeaderError: boolean = false;
  private errorToThrow: grpc.ServiceError | null = null;
  private clusterView: ClusterView | null = null;

  constructor() {
    this.server = new grpc.Server();
    this.setupKvService();
    this.setupMetaService();
  }

  /**
   * Set up mock KV service handlers.
   */
  private setupKvService(): void {
    const kvService: grpc.UntypedServiceImplementation = {
      Put: (call: grpc.ServerUnaryCall<PutRequest, PutResponse>, callback: grpc.sendUnaryData<PutResponse>) => {
        this.recordCall('Put', call.request);

        if (this.errorToThrow) {
          callback(this.errorToThrow, null);
          return;
        }

        if (this.notLeaderError) {
          const error: grpc.ServiceError = {
            name: 'NOT_LEADER',
            message: `Not the leader. Try ${this.leaderId}`,
            code: grpc.status.FAILED_PRECONDITION,
            details: this.leaderId,
          };
          callback(error, null);
          return;
        }

        const key = call.request.key.toString('hex');
        this.currentIndex++;

        const version = {
          term: this.currentTerm,
          index: this.currentIndex,
        };

        this.store[key] = {
          value: call.request.value,
          version,
          ttl: call.request.ttlMs > 0 ? call.request.ttlMs : undefined,
          createdAt: Date.now(),
        };

        const response: PutResponse = {
          version,
          meta: {},
        };

        callback(null, response);
      },

      Get: (call: grpc.ServerUnaryCall<GetRequest, GetResponse>, callback: grpc.sendUnaryData<GetResponse>) => {
        this.recordCall('Get', call.request);

        if (this.errorToThrow) {
          callback(this.errorToThrow, null);
          return;
        }

        const key = call.request.key.toString('hex');
        const entry = this.store[key];

        if (!entry) {
          const response: GetResponse = {
            value: Buffer.alloc(0),
            version: undefined,
            meta: {},
          };
          callback(null, response);
          return;
        }

        // Check TTL
        if (entry.ttl && Date.now() - entry.createdAt > entry.ttl) {
          delete this.store[key];
          const response: GetResponse = {
            value: Buffer.alloc(0),
            version: undefined,
            meta: {},
          };
          callback(null, response);
          return;
        }

        const response: GetResponse = {
          value: entry.value,
          version: entry.version,
          meta: {},
        };

        callback(null, response);
      },

      Delete: (call: grpc.ServerUnaryCall<DeleteRequest, DeleteResponse>, callback: grpc.sendUnaryData<DeleteResponse>) => {
        this.recordCall('Delete', call.request);

        if (this.errorToThrow) {
          callback(this.errorToThrow, null);
          return;
        }

        if (this.notLeaderError) {
          const error: grpc.ServiceError = {
            name: 'NOT_LEADER',
            message: `Not the leader. Try ${this.leaderId}`,
            code: grpc.status.FAILED_PRECONDITION,
            details: this.leaderId,
          };
          callback(error, null);
          return;
        }

        const key = call.request.key.toString('hex');
        const existed = key in this.store;
        delete this.store[key];

        const response: DeleteResponse = {
          tombstoned: existed,
        };

        callback(null, response);
      },
    };

    this.server.addService(KvService, kvService);
  }

  /**
   * Set up mock Meta service handlers.
   */
  private setupMetaService(): void {
    const metaService: grpc.UntypedServiceImplementation = {
      WatchCluster: (call: grpc.ServerWritableStream<ClusterView, ClusterView>) => {
        this.recordCall('WatchCluster', call.request);

        // Send initial cluster view
        if (this.clusterView) {
          call.write(this.clusterView);
        } else {
          // Default cluster view
          const defaultView: ClusterView = {
            epoch: 1,
            nodes: [
              { id: 'node-1', addr: `localhost:${this.port}`, role: 'leader' },
            ],
            shards: [],
          };
          call.write(defaultView);
        }

        // Keep stream open until client cancels
        call.on('cancelled', () => {
          call.end();
        });
      },
    };

    this.server.addService(MetaService, metaService);
  }

  /**
   * Start the mock server on a random available port.
   */
  async start(): Promise<number> {
    return new Promise((resolve, reject) => {
      this.server.bindAsync(
        '127.0.0.1:0', // Port 0 = random available port
        grpc.ServerCredentials.createInsecure(),
        (err, port) => {
          if (err) {
            reject(err);
          } else {
            this.server.start();
            this.port = port;
            resolve(port);
          }
        }
      );
    });
  }

  /**
   * Stop the mock server.
   */
  async stop(): Promise<void> {
    return new Promise((resolve) => {
      this.server.tryShutdown(() => {
        resolve();
      });
    });
  }

  /**
   * Get the server address.
   */
  getAddress(): string {
    return `127.0.0.1:${this.port}`;
  }

  /**
   * Clear all stored data.
   */
  clearStore(): void {
    this.store = {};
    this.callHistory = [];
  }

  /**
   * Get a value from the store directly (for testing).
   */
  getStoreValue(key: string | Buffer): Buffer | null {
    const keyHex = typeof key === 'string' ? Buffer.from(key).toString('hex') : key.toString('hex');
    const entry = this.store[keyHex];
    return entry ? entry.value : null;
  }

  /**
   * Set mock to return NOT_LEADER error.
   */
  setNotLeader(leaderId: string = 'mock-leader-2'): void {
    this.notLeaderError = true;
    this.leaderId = leaderId;
  }

  /**
   * Clear NOT_LEADER error state.
   */
  clearNotLeader(): void {
    this.notLeaderError = false;
  }

  /**
   * Set a custom error to throw on next request.
   */
  setError(error: grpc.ServiceError): void {
    this.errorToThrow = error;
  }

  /**
   * Clear custom error.
   */
  clearError(): void {
    this.errorToThrow = null;
  }

  /**
   * Set the cluster view to return in WatchCluster.
   */
  setClusterView(view: ClusterView): void {
    this.clusterView = view;
  }

  /**
   * Get call history for assertions.
   */
  getCallHistory(): CallRecord[] {
    return [...this.callHistory];
  }

  /**
   * Get calls for a specific method.
   */
  getCallsForMethod(method: string): CallRecord[] {
    return this.callHistory.filter(c => c.method === method);
  }

  /**
   * Clear call history.
   */
  clearCallHistory(): void {
    this.callHistory = [];
  }

  /**
   * Record a method call.
   */
  private recordCall(method: string, request: any): void {
    this.callHistory.push({
      method,
      request,
      timestamp: Date.now(),
    });
  }
}

/**
 * Create and start a mock server for testing.
 */
export async function createMockServer(): Promise<MockNoriKVServer> {
  const server = new MockNoriKVServer();
  await server.start();
  return server;
}
